pkg/policy/selectorcache.go

```diff
@@ -548,15 +548,20 @@ func (sc *SelectorCache) updateFQDNSelector(fqdnSelec api.FQDNSelector, identiti
 // an identical EndpointSelector has already been cached, the corresponding
 // CachedSelector is returned, otherwise one is created and added to the cache.
 func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api.FQDNSelector) (cachedSelector CachedSelector, added bool) {
-	sc.mutex.Lock()
-	defer sc.mutex.Unlock()
-
 	key := fqdnSelec.String()
+
+	// If the selector already exists, use it.
+	sc.mutex.Lock()
 	fqdnSel, exists := sc.selectors[key]
 	if exists {
+		sc.mutex.Unlock()
```

Comment:
Hm, should the `Unlock` be deferred here? Otherwise if multiple calls to `addUser` occur for the same `fqdnSel` at the same time, there would be concurrent access of the `users` map.

Reply:
Hmm yeah, looks like the selector users are also managed by the same selectorcache lock. I'll need to fix that up.

I'm not a big fan of the defer, because we will unlock then lock again, but this concern can be easily resolved one way or another.